### PR TITLE
Tune upload perf, better cleanup logic

### PR DESCRIPTION
**Pull request by nl0**
# Description

1. Limit hashing concurrency to prevent file access errors
2. Optimize upload code to reduce memory pressure (avoid creating and storing a closure for every upload)
3. Fall back to server-side hashing when it can't be done client-side

# TODO

- [x] [Changelog](CHANGELOG.md) entry

### CODE REVIEW

**Review by akarve — COMMENTED** [review]
I will test tomorrow and circle back

File: catalog/app/utils/ResourceCache.js [hunk 1]
```diff
@@ -148,6 +148,8 @@ function* handleInit({ resource, input, resolver }) {
 }
 
 function* cleanup() {
+  // TODO: refactor cleanup logic, so that the cleanup action is only dispatched
+  // when there's anything to cleanup (to avoid re-renders every 5 sec)
   while (true) {
```
Comment (akarve) [inline catalog/app/utils/ResourceCache.js:153]: 😬

**Review by akarve — APPROVED** [review]
I am currently testing this with a 6.9GB upload on Chrome / Mac OS / 2016 MacBook Pro.

1. OK for separate PR: We should bump the warning size to idk 10GB (I got a warning for being over 1GB)
2. The UI thread stalled for me as I was typing in the name field (after dragging and dropping my files) — about a 2.5 sec pause but then it was OK.
3. My uplink is 21.5 Mbps and I seem to be ~ about saturating it; which is good. I'd say I'll finish 6.9GB in about 35 min total wall time.
4. OK for separate PR: The progress optically appears to stall when the upload goes over 1GB but it's only because the number changes more slowly :) We should either always show more decimal places of precision above 1GB and/or we should show indeterminate progress
5. OK for separate PR; it would be really nice to have some concurrency knobs (if meaningful) in the catalog settings so that different machines/users have some control and we can learn from the field how to optimize performance

**Comment by nl0** [comment]
> The UI thread stalled for me as I was typing in the name field (after dragging and dropping my files) — about a 2.5 sec pause but then it was OK.

it was processing the dropped files (synchronously), so it's kinda expected, tho not very nice, i guess

**Comment by nl0** [comment]
> OK for separate PR; it would be really nice to have some concurrency knobs (if meaningful) in the catalog settings so that different machines/users have some control and we can learn from the field how to optimize performance

@akarve i'm not really sure how to approach this, ping me in slack for a discussion if you feel this is important

**Pull request merged**